// clashofclans.js/dist/struct/ClanMember.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClanMember = void 0;
const Constants_1 = require("../util/Constants");
const League_1 = require("./League");
class ClanMember {
    constructor(client, data) {
        this.client = client;
        this.name = data.name;
        this.tag = data.tag;
        // @ts-expect-error
        this.role = data.role.replace('admin', 'elder');
        this.expLevel = data.expLevel;
        // eslint-disable-next-line
        this.league = new League_1.League(data.league ?? Constants_1.UNRANKED_LEAGUE_DATA);
        this.trophies = data.trophies;
        this.versusTrophies = data.versusTrophies ?? null;
        this.clanRank = data.clanRank;
        this.previousClanRank = data.previousClanRank;
        this.donations = data.donations;
        this.received = data.donationsReceived;
    }
    /** Fetch detailed clan info for the member's clan. */
    async fetch(options) {
        return this.client.getPlayer(this.tag, options);
    }
}
exports.ClanMember = ClanMember;
